File: engine/engine-model/src/main/scala/client/BuddyList.scala
package client

import scala.collection.mutable
import scala.jdk.CollectionConverters._

class BuddyList(var capacity: Int) {
  private val buddies: mutable.Map[Int, BuddyListEntry] = new mutable.LinkedHashMap[Int, BuddyListEntry]

  private val pendingRequests: mutable.ArrayDeque[CharacterNameAndId] = new mutable.ArrayDeque[CharacterNameAndId]()

  def contains(characterId: Int): Boolean = {
    var result: Boolean = false
    buddies.synchronized {
      result = buddies.contains(characterId)
    }
    result
  }

  def containsVisible(characterId: Int): Boolean = {
    var buddyListEntry: Option[BuddyListEntry] = Option.empty
    buddies.synchronized {
      buddyListEntry = buddies.get(characterId)
    }
    if (buddyListEntry.isEmpty) {
      return false
    }
    buddyListEntry.get.visible
  }

  def get(characterId: Int): BuddyListEntry = {
    var buddyListEntry: Option[BuddyListEntry] = Option.empty
    buddies.synchronized {
      buddyListEntry = buddies.get(characterId)
    }
    buddyListEntry.orNull
  }

  def get(characterName: String): BuddyListEntry = {
    var buddyListEntry: Option[BuddyListEntry] = Option.empty
    buddies.synchronized {
      buddyListEntry = buddies.values.to(LazyList)
        .find(_.name.toLowerCase.eq(characterName.toLowerCase()))
    }

    buddyListEntry.orNull
  }

  def put(entry: BuddyListEntry): Unit = {
    buddies.synchronized {
      buddies.put(entry.characterId, entry)
    }
  }

  def remove(characterId: Int): Unit = {
    buddies.synchronized {
      buddies.remove(characterId)
    }
  }

  def getBuddies: java.util.Collection[BuddyListEntry] = {
    var result: java.util.Collection[BuddyListEntry] = null
    buddies.synchronized {
      result = buddies.values.asJavaCollection
    }
    result
  }

  def isFull: Boolean = {
    var result: Boolean = false
    buddies.synchronized {
      result = buddies.size >= capacity
    }
    result
  }

  def getBuddyIds: Array[Int] = {
    var result: Array[Int] = null
    buddies.synchronized {
      result = new Array[Int](buddies.size)
      result = buddies.values.to(LazyList).map(_.characterId).toArray
    }
    result
  }

  def addRequest(buddy: CharacterNameAndId): Unit = {
    pendingRequests.addOne(buddy)
  }

  def pollPendingRequest: Option[CharacterNameAndId] = {
    if (pendingRequests.isEmpty) {
      Option.empty
    } else {
      Option.apply(pendingRequests.removeLast(true))
    }
  }

  def hasPendingRequest: Boolean = {
    pendingRequests.nonEmpty
  }
}

File: engine/engine-model/src/main/java/client/BuddyList.java
package client;

import java.util.ArrayDeque;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import tools.Pair;

public record BuddyList(Integer capacity, Map<Integer, BuddyListEntry> buddies,
                        Deque<CharacterNameAndId> pendingRequests) {
   public BuddyList(Integer capacity) {
      this(capacity, new HashMap<>(), new ArrayDeque<>());
   }

   public Boolean contains(Integer characterId) {
      return buddies.containsKey(characterId);
   }

   public Boolean containsVisible(Integer characterId) {
      BuddyListEntry buddyListEntry = buddies.get(characterId);
      if (buddyListEntry == null) {
         return false;
      }
      return buddyListEntry.visible();
   }

   public BuddyListEntry get(Integer characterId) {
      return buddies.get(characterId);
   }

   public BuddyListEntry get(String characterName) {
      return buddies.values().stream().filter(buddy -> buddy.name().equals(characterName)).findFirst().orElse(null);
   }

   public BuddyList put(BuddyListEntry entry) {
      Map<Integer, BuddyListEntry> newBuddies = new HashMap<>(buddies);
      newBuddies.put(entry.characterId(), entry);
      return new BuddyList(capacity, Collections.unmodifiableMap(newBuddies), pendingRequests);
   }

   public BuddyList remove(int characterId) {
      Map<Integer, BuddyListEntry> newBuddies = new HashMap<>(buddies);
      newBuddies.remove(characterId);
      return new BuddyList(capacity, Collections.unmodifiableMap(newBuddies), pendingRequests);
   }

   public Boolean isFull() {
      return buddies.size() >= capacity;
   }

   public int[] buddyIds() {
      return buddies.values().stream().map(BuddyListEntry::characterId).mapToInt(i -> i).toArray();
   }

   public BuddyList addRequest(CharacterNameAndId buddy) {
      Deque<CharacterNameAndId> newRequests = new ArrayDeque<>(pendingRequests);
      newRequests.add(buddy);
      return new BuddyList(capacity, buddies, newRequests);
   }

   public Optional<Pair<BuddyList, CharacterNameAndId>> pollPendingRequest() {
      if (!hasPendingRequest()) {
         return Optional.empty();
      } else {
         CharacterNameAndId request = pendingRequests.removeLast();
         return Optional.of(new Pair<>(new BuddyList(capacity, buddies, pendingRequests), request));
      }
   }

   public Boolean hasPendingRequest() {
      return !pendingRequests.isEmpty();
   }

   public BuddyList updateCapacity(int capacity) {
      return new BuddyList(capacity, buddies, pendingRequests);
   }
}
